src/coreclr/src/tools/Common/TypeSystem/Common/CastingHelper.cs

```diff
@@ -20,6 +20,79 @@ public static bool CanCastTo(this TypeDesc thisType, TypeDesc otherType)
             return thisType.CanCastToInternal(otherType, null);
         }
 
+        /// <summary>
+        /// Get TypeFlags of the reduced type of a type.
+        /// The reduced type concept is described in ECMA 335 chapter I.8.7
+        /// </summary>
+        static TypeFlags GetReducedTypeElementType(TypeDesc type)
+        {
+            TypeFlags elemType = type.GetTypeFlags(TypeFlags.CategoryMask);
+            switch (elemType)
+            {
+                case TypeFlags.Byte:
+                    return TypeFlags.SByte;
+                case TypeFlags.UInt16:
+                    return TypeFlags.Int16;
+                case TypeFlags.UInt32:
+                    return TypeFlags.Int32;
+                case TypeFlags.UInt64:
+                    return TypeFlags.Int64;
+                case TypeFlags.UIntPtr:
+                    return TypeFlags.IntPtr;
+            }
+
+            return elemType;
+        }
+
+        /// <summary>
+        /// Get CorElementType of the verification type of a type.
+        /// The verification type concepts is described in ECMA 335 chapter I.8.7
+        /// </summary>
+        static TypeFlags GetVerificationTypeElementType(TypeDesc type)
+        {
+            TypeFlags reducedTypeElementType = GetReducedTypeElementType(type);
+
+            switch (reducedTypeElementType)
+            {
+                case TypeFlags.Boolean:
+                    return TypeFlags.SByte;
+                case TypeFlags.Char:
+                    return TypeFlags.Int16;
+            }
+
+            return reducedTypeElementType;
+        }
+
+        /// <summary>
+        /// Check if verification types of two types are equal
+        /// </summary>
+        static bool AreVerificationTypesEqual(TypeDesc type1, TypeDesc type2)
+        {
+            if (type1 == type2)
+            {
+                return true;
+            }
+
+            TypeFlags e1 = GetVerificationTypeElementType(type1);
+            TypeFlags e2 = GetVerificationTypeElementType(type2);
+
+            return type1.IsPrimitive && (e1 == e2);
+        }
+
+        /// <summary>
+        /// Check if signatures of two function pointers are compatible
+        /// Note - this is a simplified version of what's described in the ECMA spec and it considers
+        /// pointers to be method-signature-compatible-with only if the signatures are the same.
+        /// </summary>
+        static bool IsMethodSignatureCompatibleWith(TypeDesc fn1Ttype, TypeDesc fn2Type)
```

Comment:
I would just inline this check into the place that calls this. It's rather odd to see an API that talks about method signature compatibility taking two `TypeDesc` parameters.

Reply:
I would like to keep it in a separate method so that its name and comment can express the relation to the ECMA spec. And also match the runtime code where we actually have a separate function pointer type.

Reply:
I would at least add an assert that both types `.IsFunctionPointer`... but take this as a nit since this is a private method and I don't care that much.

Basically, I find it odd that I can do `IsMethodSignatureCompatibleWith(context.GetWellKnownType(WellKnownType.String), context.GetWellKnownType(WellKnownType.String))` and this method will return `true`.